helper_crates/document-features/lib.rs

```diff
@@ -0,0 +1,327 @@
+```
+
+## Documentation format:
+
+The documentation of the features itself goes in the `Cargo.toml`.
+
+Only the portion in the `[features]` section will be analyzed.
+Just like rust has documentation comments like `///` and `//!`, the macro will
+understands the comments that start with `## ` and `#! `. Note that the space
+is important. Lines starting with `###` will not be understood as doc comment.
+
+`## ` comments are meant to be *over* a feature and document that feature.
+there can be several `## ` comments, but they must always be followed by a
+feature name, and no other `#! ` comments in between.
+
+`#! ` comments are not associated with a particular feature, and will be printed
+in where they occurs. They are useful to do some grouping for example
+
+## Examples:
+
+*/
+// Note: because rustdoc escapes the first `#` of a line starting with `#`,
+// these docs comments have one more `#` ,
+#![doc = self_test!(/**
+[package]
+name = "..."
+## ...
+
+[features]
+default = ["foo"]
+##! This comments goes on top
+
+### The foo feature is enabling the `foo` functions
+foo = []
+
+### The bar feature enable the bar module
+bar = []
+
+##! ### Experimental features
+##! The following features are experimental
+
+### Enable the fusion reactor
+fusion = []
+*/
+=>
+    /**
+### Feature flags
+
+This comments goes on top
+* **`foo`** *(enabled by default)* —  The foo feature is enabling the `foo` functions
+* **`bar`** —  The bar feature enable the bar module
+#### Experimental features
+The following features are experimental
+* **`fusion`** —  Enable the fusion reactor
+*/
+)]
+
+extern crate proc_macro;
+use proc_macro::TokenStream;
+
+use std::collections::HashSet;
+use std::fmt::Write;
+use std::path::Path;
+use std::str::FromStr;
+
+fn error(e: &str) -> TokenStream {
+    TokenStream::from_str(&format!("::core::compile_error!{{\"{}\"}}", e.escape_default())).unwrap()
+}
+
+#[proc_macro]
+pub fn document_features(_: TokenStream) -> TokenStream {
+    document_features_impl().unwrap_or_else(std::convert::identity)
+}
+
+fn document_features_impl() -> Result<TokenStream, TokenStream> {
+    let path = std::env::var("CARGO_MANIFEST_DIR").unwrap();
+    // First try "Cargo.toml.orig", because on crates.io, Cargo.toml is usually striped
+    // of all comments, but the original Cargo.toml has been renamed
+    let cargo_toml = std::fs::read_to_string(Path::new(&path).join("Cargo.toml.orig"))
+        .or_else(|_| std::fs::read_to_string(Path::new(&path).join("Cargo.toml")))
+        .map_err(|e| error(&format!("Can't open Cargo.toml: {:?}", e)))?;
+
+    let result = process_toml(&cargo_toml).map_err(|e| error(&e))?;
+    Ok(std::iter::once(proc_macro::TokenTree::from(proc_macro::Literal::string(&result))).collect())
+}
+
+fn process_toml(cargo_toml: &str) -> Result<String, String> {
+    // Get all lines between the "[features]" and the next block
+    let lines = cargo_toml
+        .lines()
+        .skip_while(|l| l.trim() != "[features]")
```

Comment:
Is `[features] # List features here!` valid in toml? If yes, then you might want to use `starts_with`.

Reply:
good point!